bzzeth/peer.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package bzzeth
+
+import (
+	"math/rand"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// Peer extends p2p/protocols Peer and represents a conrete protocol connection
+type Peer struct {
+	*protocols.Peer            // embeds protocols.Peer
+	serveHeaders    bool       // if the remote serves headers
+	requests        *requests  // per-peer pool of open requests
+	logger          log.Logger // custom logger for peer
+}
+
+// NewPeer is the constructor for Peer
+func NewPeer(peer *protocols.Peer) *Peer {
+	return &Peer{
+		Peer:     peer,
+		requests: newRequests(),
+		logger:   log.New("peer", peer.ID()),
+	}
+}
+
+// peers represents the bzzeth specific peer pool
+type peers struct {
```

Comment:
while i can understand this approach for `peers` and `requests` i don't think that we should create a new type and a new mutex for every collection we have in the codebase. one lock will probably do and `sync.Map` will probably do the same with less code if you really insist

Reply:
i think it is cleaner code and makes reading the code easier

Reply:
I think that peers is good to have here. It provides a clearer view on its responsibilities, especially as it contains one non trivial method getEth.